File: rust/day25_part1_2021.rs

use std::fs;

fn main() {
    let contents = fs::read_to_string("input.txt").expect("Unable to read file");
    let mut grid: Vec<Vec<char>> = contents.lines().map(|line| line.chars().collect()).collect();

    let rows = grid.len();
    let cols = grid[0].len();

    let mut steps = 0;
    loop {
        steps += 1;
        let mut moved = false;
        let mut new_grid = grid.clone();

        // Move east-facing cucumbers
        for r in 0..rows {
            for c in 0..cols {
                if grid[r][c] == '>' {
                    let next_c = (c + 1) % cols;
                    if grid[r][next_c] == '.' {
                        new_grid[r][c] = '.';
                        new_grid[r][next_c] = '>';
                        moved = true;
                    }
                }
            }
        }
        grid = new_grid.clone();
        
        // Move south-facing cucumbers
        for r in 0..rows {
            for c in 0..cols {
                if grid[r][c] == 'v' {
                    let next_r = (r + 1) % rows;
                    if grid[next_r][c] == '.' {
                        new_grid[r][c] = '.';
                        new_grid[next_r][c] = 'v';
                         moved = true;
                    }
                }
            }
        }
       
        grid = new_grid;

        if !moved {
            println!("{}", steps);
            break;
        }
    }
}

File: objectivec/day25_part1_2021.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *content = [NSString stringWithContentsOfFile:@"input.txt"
                                                       encoding:NSUTF8StringEncoding
                                                          error:nil];
        NSArray *lines = [content componentsSeparatedByString:@"\n"];
        
        NSMutableArray<NSMutableString*> *grid = [NSMutableArray array];
        for (NSString *line in lines) {
            if (line.length) [grid addObject:[line mutableCopy]];
        }
        
        int steps = 0;
        int rows = (int)grid.count;
        int cols = rows ? (int)[grid[0] length] : 0;
        
        while (1) {
            BOOL moved = NO;
            NSMutableArray<NSMutableString*> *next = [NSMutableArray array];
            for (NSMutableString *row in grid) [next addObject:[[NSMutableString alloc] initWithString:row]];
            
            // East
            for (int r = 0; r < rows; r++) {
                for (int c = 0; c < cols; c++) {
                    if ([grid[r] characterAtIndex:c] == '>') {
                        int nc = (c + 1) % cols;
                        if ([grid[r] characterAtIndex:nc] == '.') {
                            [next[r] replaceCharactersInRange:NSMakeRange(c, 1) withString:@"."];
                            [next[r] replaceCharactersInRange:NSMakeRange(nc, 1) withString:@">"];
                            moved = YES;
                        }
                    }
                }
            }
            grid = next;
            next = [NSMutableArray array];
            for (NSMutableString *row in grid) [next addObject:[[NSMutableString alloc] initWithString:row]];
            
            // South
            for (int c = 0; c < cols; c++) {
                for (int r = 0; r < rows; r++) {
                    if ([grid[r] characterAtIndex:c] == 'v') {
                        int nr = (r + 1) % rows;
                        if ([grid[nr] characterAtIndex:c] == '.') {
                            [next[r] replaceCharactersInRange:NSMakeRange(c, 1) withString:@"."];
                            [next[nr] replaceCharactersInRange:NSMakeRange(c, 1) withString:@"v"];
                            moved = YES;
                        }
                    }
                }
            }
            grid = next;
            steps++;
            if (!moved) break;
        }
        printf("%d\n", steps);
    }
    return 0;
}